Add tx fee rate api methods:  gettxfeerate, settxfeerate, unsettxfeerate

ghubstan opened the pull request:
These new api methods are for viewing the `FeeService's` current tx fee rate, and setting/unsetting the custom tx fee rate preference.

This is the 21st in a chain of PRs beginning with https://github.com/bisq-network/bisq/pull/4793.  

PR https://github.com/bisq-network/bisq/pull/4855 should be reviewed before this one.

ghubstan committed:
Define new grpc bsq/btc balances protos

This change adds proto serivces and messages to support new api
implementations for serving bsq, btc or all wallet balances.

- RPC GetBsqBalances will return complete BSQ balance info.
- Message BsqBalanceInfo is proto returned by rpc GetBsqBalances.

- RPC GetBtcBalances wil return complete BTC balance info.
- Message BtcBalanceInfo is proto returned by rpc GetBtcBalances.

- RPC GetBalances returns complete BTC and BSQ balance info.
- Message BalancesInfo is proto returned by rpc GetBalances.

RPC GetBalance remains unchnaged, still returns only the available
BTC balance.  It may be deprecated and removed in a future PR.

ghubstan committed:
Add proto wrappers for serving bsq, btc or all balances

This change adds proto wrappers for sending bsq, btc, or all balances
to the CLI.  They will be used in future api method implementations:
getbsqbalance, getbtcbalance and getbalances.

ghubstan committed:
Define proto for api method 'getunusedbsqaddress'

This change adds a proto to support a future api implementation for
getting an unused, bsq funding address.

ghubstan committed:
Define proto fapi method 'sendbsq'

This change adds the proto to support a future api implementation for
transfering bsq.

ghubstan committed:
Add proto fields to support trade fee currency choice

This change adds proto support for paying maker fees in bsq or btc.
The implementation will come in a future PR.

- Added makerFeeCurrencyCode field to CreateOfferRequest proto.

- Added isCurrencyForMakerFeeBtc field to OfferInfo proto wrapper.

- Add takerFeeCurrencyCode field to TakeOfferRequest proto.

ghubstan committed:
Implement and test api method 'getunusedbsqaddress'

- Added new method to CLI, split some long msg strings into two lines,
  and added a white space after a braceless else statement.

- Added the gRPC server boilerplate.

- Added the core implementation.

- Added a test, and moved method wallet tests into their own package.

ghubstan committed:
Add teardown to test case

ghubstan committed:
Remove trailing spaces in blank line

ghubstan committed:
Implement and test new getbalance(s) api methods

- Added three new methods to CLI:

      getbalances   ...	returns complete bsq and btc balance info
      getbsqbalance ...	returns complete bsq balance info
      getbtcbalance ...	returns complete btc balance info

      The old getbalance method is deprecated and will be removed
      if there is agreement to do that.

- Made the needed changes in the CLI's output formatting classes.

- Added new tests to existing BsqWalletTest, added new BtcWalletTest
  and WalletBalancesTest.

- Added disabled tests for funding a bsq wallet (todo in next PR).

ghubstan committed:
Resolve unnecessary use of fully qualified name for codacy

ghubstan committed:
Stub out api method 'sendbsq' in core

All the boilerplate for this method is included in this change, but
the implementation in CoreWalletsService#sendBsq is commented out
until the needed logic to send bsq is refactored out of desktop's
BsqSendView class -- to be included in the next PR.

- Added new method to CLI.

- Added the gRPC server boilerplate.

- Added the core implementation, commented out.

- Enabled tests.

ghubstan committed:
Re comment sendbsq tests so travis ci does not fail

ghubstan committed:
Refactor desktop's BsqSendView, share with api

Moved just enough code out of BsqSendView to avoid desktop/api
'sendbsq' duplication, at the cost of adding 1 new method to
BsqSendView.

- Created new BsqTransferModel to hold tx details shared by desktop and api.

- Created new BsqTransferService to send bsq using a BsqTransferModel shared
  by desktop and api.

- Uncommented CoreWalletsService#sendBsq implementation.

- Uncommented sendbsq tests.

ghubstan committed:
Resolve 'Avoid creating BigDecimal with a decimal' issue for codacy

ghubstan committed:
Support paying trade fees in bsq or btc (api)

- Add optional makerFeeCurrencyCode argument to CLI's 'createoffer'.

- Add optional takerFeeCurrencyCode argument to CLI's 'takeoffer'.

- Add isCurrencyForMakerFeeBtc field to OfferInfo proto wrapper.

- Pass fee currency argument down to CoreOffersService and CoreTradesService's
  create and take offer methods.

- Add maybeSetFeePaymentCurrencyPreference method to OfferUtil to
  conditionally set the 'payFeeInBtc' preference before creating
  or taking an offer.

- Adjust affected tests.

ghubstan committed:
Delete deprecated api test, adjust api build/run doc

ghubstan committed:
Refactor api getbalance methods.

This change fixes the recently bloated wallet balances api.  Now there
is one CLI getbalance [bsq|btc] method that calls a getbalances(currency-code)
on the server.  The server returns full wallet balance information for
BSQ, BTC, or both if the CLI's currency argument is absent.

- grpc.proto:  Reduced number of getbalance(s) proto services from
  four to one.

- GrpcWalletsService:  Reduced number of getbalance(s) gRPC service
  boilerplate methods from four to one.

- CoreApi, CoreWalletsService:  Reduced number of getbalance(s) gRPC
  service implementation methods from four to one.

- CliMain:  Reduced number of getbalance(s) commands from four to one.

- BalancesInfo: Changed BsqBalanceInfo & BtcBalanceInfo field names
  to bsq and btc, respectively, to make calling their accessors
  more readable, less verbose.

- BtcBalanceInfo, BsqBalanceInfo: Defined static final EMPTY instances
  as place holders in a BalancesInfo returned by the gRPC server, when
  one or the other balance types is not requested.  Would be nice
  to serve Optional.empty() instead, but protobuf does not support
  it or null.

- Adjusted affected api tests and build doc.

ghubstan committed:
Fix typo

ghubstan committed:
Remove unnecessary fully qualified name

ghubstan committed:
Remove unused imports

ghubstan committed:
Add api method 'getpaymentmethods'

Returns a list of supported payment method IDs.  This gives CLI users
the correct payment method id for creating a new payment account.

- Added gRPC service GetPaymentMethods to grpc.proto.

- Added gRPC boilerplate method to GrpcPaymentAccountsService.

- Added implimentation to CoreApi -> CorePaymentAccountsService.

- Refactored PaymentAccountTest hierarchy.

- Add api method to CLI.

ghubstan committed:
Avoid codacy issue over use of fully qualified name

Had to change the getPaymentMethods() names to getPaymentMethodIds() to
avoid this codacy issue:

"Unnecessary use of fully qualified name 'PaymentMethod.getPaymentMethods'
due to existing static import 'bisq.core.payment.payload.PaymentMethod.*'"

If 'PaymentMethod.getPaymentMethods' was changed to 'getPaymentMethods',
a recursive loop would result, ending in an out of stack heap crash.

This renaming of the method on the server is correct, but the CLI's
'getpaymentmethods' name was not changed.

ghubstan committed:
Revert "Avoid codacy issue over use of fully qualified name"

This reverts commit a46526198dc4667621d0d3bc6f5b86d7a4f2d6a3.

The service side method name change is not correct.

ghubstan committed:
Print the payment method id (only)

ghubstan committed:
Avoid codacy issue over use of fully qualified name

Changed the core getPaymentMethods() -> getFiatPaymentMethods() to avoid

"Unnecessary use of fully qualified name 'PaymentMethod.getPaymentMethods'
due to existing static import 'bisq.core.payment.payload.PaymentMethod.*'"

ghubstan committed:
Make codacy happy again

ghubstan committed:
Add ReflectionUtils to common.util pkg

This class will aid the api's (create) PaymentAccount json form
serialization/de-serialization.

ghubstan committed:
Add new (gson) PaymentAccountTypeAdapter to core.api.model

This class does most of the work of the api's (create) PaymentAccount
json form serialization/de-serialization.

ghubstan committed:
Add isCountryBasedPaymentAccount to abstract PaymentAccount

ghubstan committed:
Add new PaymentAccountForm to core.api.model

This class uses the gson PaymentAccountTypeAdapter to generate editable
json forms for PaymentAccount sub-types, and instantiate PaymentAccount
instances from edited json forms.

ghubstan committed:
Add boilerplate for new 'getpaymentacctform' api method

ghubstan committed:
Add new api method 'getpaymentacctform' to CLI

ghubstan committed:
Test new api method 'getpaymentacctform'

ghubstan committed:
Ensure EXPECTED_FORM.clear() is never skipped

ghubstan committed:
Replace existing api method 'createpaymentacct' impl

In the CLI, this method now takes a single argument, the path to a json
file (a payment account form).  The json text is read from the file
and sent to the server, where it is serialized to a new PaymentAccount
instance, saved, and sent back to the CLI as a protobuf.PaymentAccount.

- Modified existing gRPC boilerplate in grpc.proto and GrpcPaymentAccountsService.

- Modified existing method signatures in CoreApi and CorePaymentAccountsService.

- Replaced existing method impl in CorePaymentAccountsService and
  removed dead code.

- Replaced the CLI's existing 'createpaymentacct' method impl.

- Modified existing tests.

ghubstan committed:
Avoid 'unnecessary use of fully qualified name' codacy issue

ghubstan committed:
Tidy up payment accts api related msgs on CLI

ghubstan committed:
Fix createpaymentacct validation problems

- Add missing boilerplate error handling to GrpcPaymentAccountsService.

- Edit some PaymentAccountForm & PaymentAccountTypeAdapter exception
  messages that would be passed to CLI;  they should be in the same
  style as existing CLI err msgs.

ghubstan committed:
Log server stack traces, pass concise err msgs to CLI

For exceptions caught in the server, we want to log a reason and a
full stack trace, then send less verbose exception msgs up to the CLI.

ghubstan committed:
Add create payment account (api) tests

ghubstan committed:
Reverse argument order for assert(expected, actual) convention

ghubstan committed:
Add isMoneyGramAccount to abstract PaymentAccount

ghubstan committed:
Refactor didReadCountryField to set country on MoneyGram acct

MoneyGram is not a CountryBasedPaymentAccount, but it does
have a country field.

ghubstan committed:
Enable CreateMoneyGramAccount test

ghubstan committed:
Fix broken tests

- Change the getbalance tests to just check a sucessful '0' status code.

  The api's getbalance method returns full balance info now, not just a
  formatted long.  We still assert the cmd did not fail, but don't check
  the value string.

- Remove obsolete createpaymentacct tests.

  This is well tested in apitest, and testing this cmd is unnecesarily
  complex to do in a bats script.

ghubstan committed:
Add two new api method tests

Keeping the bats test simple.  More complex use cases are tested in
apitest, and we don't want to be trading on mainnet with a bats test.

- Add new getunusedbsqaddress test, assert success return status.
- Add new getpaymentmethods test, assert success return status.

ghubstan committed:
Fix getbalance out bug in CLI

Running the bats test exposed a bug that shows BSQ and BTC balance
info no matter what currency code argument is passed to the server.

ghubstan committed:
Print 'no offers found' msg in CLI if list is empty

ghubstan committed:
Fix minor error & help msg bugs

ghubstan committed:
Resolve BsqSendView file conflict

Restore BsqSendView to version @ commit 48a515be01dd674d241be04cddb18c0e823e91b4.
Author: jmacxx
Date:   Tue Nov 10 19:35:31 2020 -0600

    Add encrypted wallet password prompt when sending funds from BSQ wallet

    Correct wording of transaction confirmation popup to use 'mining fee'
    instead of 'transaction fee' to make it consistent with wording of
    the BTC confirmation popup.

ghubstan committed:
Remove deprecated method long getAvailableBalance

ghubstan committed:
Add missing newlines to console output statements

And remove some unecessary single quotes.

ghubstan committed:
Implement fromProto

This is marked unused, but it will be in future tests.

ghubstan committed:
Merge branch 'master' into 20-pr-change-requests-A

ghubstan committed:
Adjust to changed CoinUtil

ghubstan committed:
Adjust test for a new payment method

ghubstan committed:
Add protobuf defs for gettxfeerate, settxfeerate, unsettxfeerate

ghubstan committed:
Add protbuf TxFeeRateInfo

ghubstan committed:
Add TxFeeRateInfo proto wrapper

ghubstan committed:
Add gettxfeerate, settxfeerate, unsettxfeerate implementations

ghubstan committed:
Test new tx fee rate api methods

ghubstan committed:
Remove unnecessary use of fully qualified name for codacy

sqrrm commented on cli/src/main/java/bisq/cli/CurrencyFormat.java:
```diff
@@ -50,6 +53,23 @@ public static String formatBsq(long sats) {
+        String stdTxFeeRate = formatTxFeeRate(txFeeRateInfo.getStdTxFeeRate());
+        String customTxFeeRate = txFeeRateInfo.getCustomTxFeeRate() < 0
+                ? formatTxFeeRate(txFeeRateInfo.getCustomTxFeeRate())
+                : null;
```
I don't understand what's going on here.

ghubstan replied:
To clear this up, I should probably add a `boolean useCustomWithdrawalTxFee` field to `TxFeeRateInfo`.  This would be set from the `Preference#useCustomWithdrawalTxFee` value.     (Remember that `Preferences` is not visible to the CLI, and these `*Info` proto wrappers are for passing the minimum data needed from server to CLI.)

This confusion is caused by not having that field.  When `txFeeRateInfo.getCustomTxFeeRate() == -1`, the custom tx fee rate preference is ignored in the CLI.  In core, it means the preference.useCustomWithdrawalTxFee == false.

In pseudo English, it means

```
String customTxFeeRate = (if are we using the fee service's tx fee rate)
                ? (format the custom tx fee rate)
                : null;

```

Adding a `boolean useCustomWithdrawalTxFee` field to `TxFeeRateInfo` would fix this.

ghubstan replied:
Fixed in commit faf030fbc5810e46a0d6fdab154b154ef39f72db.

sqrrm commented on core/src/main/java/bisq/core/api/CoreWalletsService.java:
```diff
@@ -185,6 +196,37 @@ void sendBsq(String address,
+    TxFeeRateInfo getTxFeeRate() {
+        try {
+            CompletableFuture<Void> feeRequestFuture = CompletableFuture.runAsync(feeService::requestFees);
+            feeRequestFuture.get();  // Block until async fee request is complete.
```
Is it really a good idea to block here? As discussed before, could we use a callback to avoid blocking?

ghubstan replied:
I'm not sure what the difference would be between future.get() and using a callback.   The wait for a response has to happen somewhere before `GrpcWalletsService` returns the gRPC reply.

Do you think `FeeService#requestFees` should be refactored to take a callback parameter?  Were you and chimp wanting to  refactor FeeService#requestFees?  Or should that be left alone, and make adjustments only in the api?

Another way would be to deal with the result handling is do it in the `GrpcWalletsService`.  I try to keep those boilerplate `Grpc*Service` classes as small as possible, but sometimes dealing with async calls in that layer is unavoidable.

If we are concerned about blocking forever, we could block with a timeout:  
`feeRequestFuture.get(5, SECONDS);  // Block until async fee request is complete.`

sqrrm replied:
I was thinking something like, in CoreWalletService
```
    TxFeeRateInfo getTxFeeRate(ResultHandler resultHandler, ErrorMessageHandler errorMessageHandler) {
            CompletableFuture<Void> feeRequestFuture = CompletableFuture.runAsync(feeService::requestFees);
            Futures.addCallback(feeRequestFuture, new FutureCallback<>() {

                @Override
                public void onSuccess(@Nullable Object result) {
                    resultHandler.handleResult();
                }

                @Override
                public void onFailure(Throwable t) {
                    errorMessageHandler.handleErrorMessage(t.getMessage());
                }
            }, MoreExecutors.directExecutor());
    }
```
where the resultHandler is doing the gRPC reply, something like
```
    @Override
    public void getTxFeeRate(GetTxFeeRateRequest req,
                             StreamObserver<GetTxFeeRateReply> responseObserver) {
        ResultHandler resultHandler = () -> {
            var reply = GetTxFeeRateReply.newBuilder()
                    .setTxFeeRateInfo(txFeeRateInfo.toProtoMessage())
                    .build();
            responseObserver.onNext(reply);
            responseObserver.onCompleted();
        };
        coreApi.getTxFeeRate(resultHandler, errorHandler);
    }
```
Bad code, just to illustrate the idea.

ghubstan replied:
This is exactly [erhmmm... similar] what I'm doing, will test soon...

ghubstan replied:
Commit 987d89319e8a824033879846ac43003ea219cfd0 is a solution to the blocking problem.  I don't pass an error handler to keep it simple, and not do that before I understand how it would help here.

sqrrm commented on core/src/main/java/bisq/core/api/CoreWalletsService.java:
```diff
@@ -185,6 +196,37 @@ void sendBsq(String address,
+    }
+
+    TxFeeRateInfo setTxFeeRatePreference(long txFeeRate) {
+        if (txFeeRate <= 0)
```
We also don't want to create txs with fees < 1 sat/b, but perhaps that doesn't have to be checked here.

ghubstan replied:
Doing this check here makes the CLI's error msg stylistically consistent with others.  

This brings up a general issue that is going to need quite a bit of work in the API:  displaying validation error messages coming from the server's core, in the CLI.    But for now, getting all the v1 methods working for correct parameters is the short term priority.

sqrrm replied:
Agreed, not an issue for now, good to think of how to handle it though as you say.

ghubstan committed:
Remove many log.isDebugEnabled() checks

These do not buy enough in the way of performance, at the cost of
so many additional lines.  Buy these checks should be done for
debug log statements that have expensive parameters.

ghubstan committed:
Don't be so polite

ghubstan committed:
Loosen up the PaymentMethod count test assert

, so we don't have to change this test every time a new payment
method is added.

ghubstan committed:
Add useCustomTxFeeRate field to TxFeeRateInfo proto wrapper

This is set from the core preferences.isUseCustomWithdrawalTxFee(),
and simplifies fee changing logic in the API.

ghubstan committed:
Use ListenableFuture and callback when requesting tx fee

This change fixes the blocking problem in the fee rate request api.
Also redefined the TxFeeRateInfo.

- Redefined grpc.proto message TxFeeRateInfo, added
  lastFeeServiceRequestTs field. (CLI user may want to know
  TS of last fee request.)

- Adjusted TxFeeRateInfo proto wrapper.

- Adjusted CurrencyFormat and BtcTxFeeRateTest to new TxFeeRateInfo.

- Added @Getter annotation to FeeService.  (CLI user may want to know
  TS of last fee request).

- Pass resultHandler from GrpcWalletsService through CoreApi, to
  CoreWalletsService's tx fee rate api methods.

sqrrm reviewed (commented):
utACK

I think this looks much better and this is probably ready for merging now. We'll have to put some more thought into the general thread handling from the api side but that's not for this PR.

ghubstan committed:
Make salt an editable payment acct json form field

Users need to be able to preserve their acct age when moving a
payment account to a new client.

Also adjusted affected tests, and did some minor refactoring
of the custom gson type adaptor.

ghubstan committed:
Change sendbsq's amount parameter type to String

ghubstan committed:
Remove method only used in class' toString

ghubstan committed:
Prevent attempt to send amount of BSQ < dust

sqrrm reviewed (approved):
utACK

ghubstan commented:
Force another travis build because a dependency download failed that last time.
